utils/proc/process.go

```diff
@@ -0,0 +1,208 @@
+	return
+}
+
+type ps struct {
+	imp *process.Process
+}
+
+func (p *ps) Cmdline() string {
+	cmd, _ := p.imp.Cmdline()
+	return cmd
+}
+
+func (p *ps) Cwd() string {
+	cwd, _ := p.imp.Cwd()
+	return cwd
+}
+
+func (p *ps) Parent() IProcess {
+	pp, _ := p.imp.Parent()
+	return wrapProcess(pp)
+}
+
+func (p *ps) Children(ctx context.Context) (children []IProcess, err error) {
+	err = parallelisation.DetermineContextError(ctx)
+	if err != nil {
+		return
+	}
+	cp, err := p.imp.Children()
+	err = ConvertProcessError(err)
+	if err != nil {
+		return
+	}
+	for i := range cp {
+		children = append(children, wrapProcess(cp[i]))
+	}
+	return
+}
+
+func (p *ps) Name() string {
+	n, _ := p.imp.Name()
+	return n
+}
+
+func (p *ps) Environ(ctx context.Context) (env []string, err error) {
+	env, err = p.imp.EnvironWithContext(ctx)
+	err = ConvertProcessError(err)
+	return
+}
+
+func (p *ps) Pid() int {
+	return int(p.imp.Pid)
+}
+
+func (p *ps) PPid() int {
+	ppid, _ := p.imp.Ppid()
+	return int(ppid)
+}
+
+func (p *ps) Executable() string {
+	ex, _ := p.imp.Exe()
+	return ex
+}
+
+func (p *ps) Terminate(ctx context.Context) error {
+	err := ConvertProcessError(p.imp.TerminateWithContext(ctx))
+	if commonerrors.Any(err, commonerrors.ErrNotFound) {
+		err = nil
+	}
+	return err
+}
+
+func (p *ps) KillWithChildren(ctx context.Context) error {
+	return killProcessAndChildren(ctx, p.imp)
+}
+
+func killProcessAndChildren(ctx context.Context, p *process.Process) (err error) {
+	err = parallelisation.DetermineContextError(ctx)
+	if err != nil {
+		return
+	}
+	if p == nil {
+		return
+	}
+	defer func() {
+		_ = p.Kill()
+	}()
+	err = ConvertProcessError(p.TerminateWithContext(ctx))
+	if err != nil {
+		if commonerrors.Any(err, commonerrors.ErrNotFound) {
+			err = nil
+		}
+		return
+	}
+	err = ConvertProcessError(killGroup(ctx, p.Pid))
+	if err == nil {
+		if commonerrors.Any(err, commonerrors.ErrNotFound) {
+			err = nil
+		}
+		return
+	}
+	err = killChildren(ctx, p)
+	if err != nil {
+		if commonerrors.Any(err, commonerrors.ErrNotFound) {
+			err = nil
+		}
+		return
+	}
+	err = ConvertProcessError(p.KillWithContext(ctx))
+	if commonerrors.Any(err, commonerrors.ErrNotFound) {
+		err = nil
+	}
+	return
+}
+
+func killChildren(ctx context.Context, p *process.Process) (err error) {
+	err = parallelisation.DetermineContextError(ctx)
+	if err != nil {
+		return
+	}
+	if p == nil {
+		return
+	}
+	children, childErr := p.ChildrenWithContext(ctx)
+	if childErr != nil {
+		childErr = ConvertProcessError(childErr)
+		if commonerrors.Any(childErr, commonerrors.ErrTimeout, commonerrors.ErrCancelled) {
+			err = childErr
+		}
+		return
+	}
+	for i := range children {
+		subErr := killProcessAndChildren(ctx, children[i])
+		if subErr != nil {
+			err = subErr
+		}
+	}
+	if err != nil {
+		_ = killGroup(ctx, p.Pid) // Radical approach of killing the whole process group
+	}
+	return err
```

Comment:
do you still want to return the error from before the `killGroup`? This will mean the function errors even if `killGroup` succeeds

Reply:
fair point if no error is returned I will set err to nil